utils/docs/gschemdoc.1.in
.TH gschemdoc 1 "@DATE@" "gEDA Project" @VERSION@
.SH NAME
gschemdoc - Present relevant documentation
.SH SYNOPSIS
\fBgschemdoc\fR [\fIOPTION\fR] \fIDOCUMENTATION\fR [\fIDEVICE\fR \fIVALUE\fR \fISYMBOL\fR \fIDIRECTORY\fR]
or
\fBgschemdoc\fR \fIOPTION\fR \fIPATH\fR

.SH DESCRIPTION
.PP
\fBgschemdoc\fR is shell script that attempts to present relevant information
or documentation with regard to a specifc component, manual, or URL. The order
of arguments sets the presidence used by \fBgschemdoc\fR, with the first argument,
\fIDOCUMENTATION\fR, having the highest priority.
.PP
The \fIDOCUMENTATION\fR argument can be a URL, either local or remote, or a PDF
file. Without an option, the \fIGEDADOCDIR\fR directory is checked first, if a
file is not found and a BROWSER is found then the \fIDOCUMENTATION\fR is passed
to the browser. If the \fIDOCUMENTATION\fR argument is empty, the scripts checks
for DEVICE-VALUE.pdf and then a DEVICE.pdf in the \fIGEDADOCDIR\fR directory,
followed by browsing for "DEVICE VALUE" if a BROWSER was found and the \fIVALUE\fR
argument is present, otherwise only browsing for "DEVICE". If a BROWSER was
not found, then the scripts attempts to locate a pdf file in \fIGEDADOCDIR\fR
directory with the base name of \fISYMBOL\fR.

.SH OPTIONS
.TP 8
\fB-m\fR
Display user's manual instead; Using this option loads "gedadocs.html" in a
browser, all other arguments are ignored.
.TP 8
\fB-w\fR
Display a wiki page. When this option is given, the script attempts to show
the local wiki page for \fIDEVICE\fR.

.SH AUTHORS
See the `AUTHORS' file included with this program.

.SH COPYRIGHT
.nf
Copyright \(co 1999-@YEAR@ gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
\fBgaf\fR(1), \fBgschem\fR(1)


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2017 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 11/12/17 Rampage Edition

* Expanded support for internationalization (i18n).

* The build system now support using the --with-included-gettext option to
  configure the suite to build and use an in-source international library.

* Font-sizes for widgets on GedaToolbars can now be specified in RC files.

* The toolbar layout in gschem has been improved to better support labels.

* The visibility of toolbar tooltips can now be control using RC keywords.

* The Slot Editor Dialog now incorporates a spinner control to adjust the
  slot number of slotted components. The range is restricted to valid slots
  for the selected component.

* New bitmaps have been added;

      * geda-deslect.png
      * geda-find-attribute
      * geda-properties.png
      * geda-trash-can.png

* libgedathon no longer over-writes existing gafrc when the declare_local_sym
  feature is used.

* libgedathon now supports internationalization (i18n).

Notable changes in gEDA/gaf gEDA 2.1.1.0 - 01/01/17 Rampage Edition

* Improvements were made to both the build system and source code to ensure
  gEDA-GAF builds more reliably on 64 bit platforms. Testing was performed
  using Ubuntu 12.04 and Ubuntu 14.04 64 bit host systems.

* The libgeda API has been completely redone in order to present a library API
  that is not only for in-suite applications.

* Support for internationalization (i18n) has been significantly expanded and
  now includes command-line utilities.

* Extensive revisions have been made to improve performance of gEDA application.
  This is particularly noticeable in gschem.

* A new library, libgedacolor, now unifies color management for the gEDA suite.
  Color-maps are still configured using scheme files, but the use of "friendly-
  color-maps"  is no longer applicable; the new color maps for libgedacolor are
  already "friendly". The library was added to provide a unified API for gEDA
  applications and should be transparent to most users. User who are using
  custom color-maps will need to modify their custom map files.

* A new auto-panning feature has been added to gschem. The auto-pan is only in
  effect while drawing nets and buses and can enabled or disabled through menu
  options.

* The Auto-Number dialog in gschem no longer requires the use of wild-card
  characters in the scope text. This is a new advanced feature that allows
  automatic re-numbering of duplicated reference designators.

* The path for logs file can now be set in initialization files, including
  local gafrc files. This allows logs to be organized on a per project basis.
  The location for logs can also be set using a new GEDALOGS environment variable.

* A new libgeda keyword "log-time" can be use to enable or disable prefixing
  log entries with the time of day.

* The Multi-Attribute Editor Dialog entry now has attribute "word" completion
  identical to the Single-Attribute Editor Dialog.

* The list of valid attribute used by gsymcheck when checking symbol files is
  now users configurable. The list can be set in system-gsymcheckrc.

* The list of known devices recognized by gsymcheck when checking the device
  entries in symbol files is now users configurable. The list can be set in
  system-gsymcheckrc.

* Symbol files can now be renamed using the Component Dialog.

* The number of user defined colors was increased by 6 to a total of
  10 user definable colors in each color map.

* The main menu system has been modified to automatically build a basic
  fall-back menu when errors are discovered in the menu RC file.

* The height of the status bar in Gschem can now be set in the gschem-gtkrc

* HotKey dialog now includes actions icon and a column to display the actual
  binding string. User can enable or disable viewing the action column with
  the setting being retained between session.

* The Slot Editor Dialog now displays the total number of slots per component.

* New features have been added to the Page Select Dialog:

      Accept drag-and-drop of schematic or symbol files.
      Opened pages can be re-ordered either up or down.
      Auto scrolls to selection.
      Shows long filename as tooltip when showing short names.
      Optionally adjust height of dialog automatically.

* Exchange image file now retains the user preference to update all selected
  images between session. Previously the user preference was only retained for
  current session.

* Additional preference variables have been incorporated in new the configuration
  system allow retention of changes using the Preferences dialog:

      continue-component-place
      embed-components
      enforce-hierarchy
      force-boundingbox
      keyboardpan-gain
      netconn-rubberband
      select-slack-pixels
      snap-size
      sort-component-library
      object-clipping

  Uses of these keywords in rc files is still permissible and will over-ride the
  preference setting.

* A new gschem command line options has been provided to specify a title-block at
  startup.

* A new gschem command line options has been provided to startup minimized.

* The command line option to add a directory to the Guile path now has a long option.
  The "-L" option has been replaced with "-g" or "--guile".

* gsymcheck now supports the standard command-line version options.

* Programs in the geda-gaf suite now support quite mode in conjunction with the version
  option to assist with extraction of dotted version information.

* New Scheme API have been added:

      gschem-bmp-image
      gschem-jpeg-image
      gschem-tiff-image

* The maximum zoom magnification in gschem has been increased slightly.


gschem/scheme/gschem/gschemdoc.scm
;; gEDA - GPL Electronic Design Automation
;; libgeda - gEDA's library - Scheme API
;;
;; Copyright (C) 2011-2015 Peter Brett <peter@peter-b.co.uk>
;;
;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 51 Franklin Street, Boston, MA 02110-1301 USA
;; <http://www.gnu.org/licenses/>
;;

(define-module (gschem gschemdoc)
  #:use-module (gschem core gettext)
  #:use-module (geda os)
  #:use-module (geda object)
  #:use-module (geda attrib)
  #:use-module (gschem util)
  #:use-module (ice-9 optargs)
  #:use-module (ice-9 regex)
  #:use-module (srfi srfi-1))

;; FIXME this should probably be set via configure
(define %configured-sys-doc-dir #f)

(define-public (sys-doc-dir)
  "sys-doc-dir

Get the directory where gEDA documentation is stored."

  ;; Given a gEDA data directory path, guess the installation prefix.
  (define (guess-prefix dir)
    (let ((ofs (string-contains
                dir (string-append separator "share"))))
      (and ofs (substring dir 0 ofs))))

  ;; Guess the gEDA documentation directory, using the gEDA data
  ;; search path.  Guaranteed to only return paths that correspond to
  ;; actual directories.
  (define (guess-docdir)
    (any

     (lambda (dir)
       (let ((docdir
              (string-join (list (guess-prefix dir)
                                 "share" "doc" "packages" "geda-gaf")
                           separator)))
         (and (false-if-exception
               (eq? 'directory (stat:type (stat docdir))))
              docdir)))

     (sys-data-dirs)))

  (or (guess-docdir)
      (if (platform? 'win32-native)
          #f
          %configured-sys-doc-dir)))

(define-public (user-doc-dir)
  "user-doc-dir

Get the directory where per-user gEDA documentation is stored."

  (string-join (list (car (user-data-dir))
                     "doc" "geda-gaf")
               separator))

;; Munge a wiki page name so that it can be used in a filename
(define (wiki-munge name)
  (string-map
   (lambda (c)
     (case c
       ((#\? #\\ #\! #\*) #\_)
       ((#\:) #\-)
       (else c)))
   name))

(define*-public (show-wiki #:optional (page "index"))
  "show-wiki PAGE

Launch a browser to display a page from the offline version of the
gEDA wiki shipped with gEDA/gaf.  The specified PAGE should be a
string containing the page name as used on the live version of the
wiki."

  (show-uri
   (string-append "file://"
                  (string-join (list (sys-doc-dir) "wiki")
                               separator 'suffix)
                  (wiki-munge page)
                  ".html")))

(define %pdf-search-template "http://www.google.com/search?q=~S%20filetype:pdf")

;; Get the value of a named attribute.  Attached attributes are
;; searched first, followed by inherited attributes.  The first
;; non-empty result found is returned.
(define (attribute-value-by-name obj name)
  (define (any-proc attrib)
    (and (equal? name (attrib-name attrib))
         (let ((v (attrib-value attrib)))
           (and (> (string-length v) 0) v))))
  (or (any any-proc (object-attribs obj))
      (any any-proc (inherited-attribs obj))))

;; For each entry in DIRNAME, PROC is called.  If PROC returns a
;; true value, the iteration stops and the result returned by PROC
;; is returned to the caller.
(define (directory-any dirname proc)
  (call/cc
   (lambda (return)
     (let ((dir #f))
       (dynamic-wind
           (lambda () (set! dir (opendir dirname)))
           (lambda ()
             (do ((entry (readdir dir) (readdir dir)))
                 ((eof-object? entry))
               (if (member entry '(".." "."))
                 #f
                 (let ((result (proc entry)))
                   (if result (return result) #f)))))
           (lambda () (closedir dir))))
     #f)))

;; Searches for and displays documentation in DIRNAME.  A
;; documentation file is expected to start with BASENAME and
;; (optionally) end with EXT.  Comparisons are carried out
;; case-insensitively.  If no file was found, returns #f.
(define* (directory-doc-search dirname basename #:optional (ext ""))
  (define (test-dir-entry entry)
    (let ((filename (string-append dirname separator entry)))
      (and (string-prefix-ci? basename entry)
           (string-suffix-ci? ext entry)
           (file-exists? filename)
           filename)))
  (let ((filename (false-if-exception (directory-any dirname test-dir-entry))))
    (and filename (begin (show-file filename) #t))))

;; Searches for documentation STRING on the Internet, using
;; %pdf-search-template.
;;
;; FIXME string should be URL-encoded.
(define (internet-doc-search string)
  (show-uri (format #f %pdf-search-template string)) #t)

;; Munges a component basename to look more like a device name
(define (munge-basename basename)
  (let* ((rx (make-regexp "(-[0-9]+)?.sym$" regexp/icase))
         (match (regexp-exec rx basename)))
    (if match
        (regexp-substitute #f match 'pre)
        basename)))

(define-public (show-component-documentation obj)
  "show-component-documentation COMPONENT

Find the documentation for COMPONENT, by inspecting (in order) its
\"documentation\", \"device\" and \"value\" attributes, and its
component basename, and searching in the current directory, the user
and system gEDA documentation directories, and Google. In most cases,
results are restricted to \".pdf\" files. The documentation is
displayed in the system associated viewer application."


  (let ((documentation (attribute-value-by-name obj "documentation"))
        (device (attribute-value-by-name obj "device"))
        (value (attribute-value-by-name obj "value"))
        (basename (component-basename obj)))

    ;; Try several different ways of finding documentation
    (or

     ;; 1) Checks based on "documentation=" attribute
     (and
      documentation

      (or
       ;; a) First check the obvious -- does the documentation
       ;;    attribute match a file in the current directory?
       (directory-doc-search (getcwd) documentation)
       ;; b) What about in the documentation directories?
       (directory-doc-search (user-doc-dir) documentation)
       (directory-doc-search (sys-doc-dir) documentation)
       ;; c) Does the documentation attribute look like a URL?
       (and (any (lambda (prefix) (string-prefix? prefix documentation))
                 '("http://" "ftp://" "file://"))
            (show-uri documentation))

       ;; d) If a documentation attribute was specified at all, search
       ;;    for it with Google.
       (internet-doc-search documentation)))

     ;; 2) Checks based on "device=" and "value=" attributes
     (and
      device value
     (let ((device-value (string-append device "-" value)))
     (or
     ;; a) Look for a DEVICE-VALUE*.PDF file in the current directory
        (directory-doc-search (getcwd) ".pdf")
     ;; b) Look for a DEVICE-VALUE*.PDF file in the documentation directories
        (directory-doc-search (user-doc-dir) device-value ".pdf")
        (directory-doc-search (sys-doc-dir) device-value ".pdf"))))
     (and
      device
      (or
       ;; c) Look for a DEVICE*.PDF file in the current directory
       (directory-doc-search (getcwd) device ".pdf")
       ;; d) Look for a DEVICE*.PDF file in the documentation directories
       (directory-doc-search (user-doc-dir) device ".pdf")
       (directory-doc-search (sys-doc-dir) device ".pdf")
       ;; d) If there's a device attribute, search for it with Google.
       (internet-doc-search device)))
     (and
      value
      ;; e) If there's a value attribute, search for it with Google.
      (internet-doc-search value))

     ;; 3) Checks based on component basename
     (and
      basename (string-length basename) ; Check basename is non-empty

      ;; Munge basename to look more like a device name
      (let ((name (munge-basename basename)))
        (or
         ;; a) Look for BASENAME*.PDF file in current directory
         (directory-doc-search (getcwd) name ".pdf")
         ;; b) Look for BASENAME*.PDF file in documentation directories
         (directory-doc-search (user-doc-dir) name ".pdf")
         (directory-doc-search (sys-doc-dir) name ".pdf"))))

     ;; 4) Fail miserably
     (error (_ "No documentation found")))))
